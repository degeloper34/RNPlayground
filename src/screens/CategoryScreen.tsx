import {Ionicons} from "@expo/vector-icons";
import {useContext, useLayoutEffect, useState} from "react";
import {FlatList, Pressable, StyleSheet, View} from "react-native";
import {Product, RootStackScreenProps} from "../../types";
import {CustomTextInput, EmptyState} from "../components/atoms";
import {ProductCard} from "../components/molecules";
import Colors from "../constants/Colors";
import {MainContext} from "../context/mainContext";

export default function CategoryScreen({
  navigation,
  route,
}: RootStackScreenProps<"Category">) {
  const {category} = route?.params;
  const [searchText, setSearchText] = useState("");
  const context = useContext(MainContext);

  const selectedCategory = context.productByCategory[category];

  const filteredCategoryList = selectedCategory?.filter((item: Product) =>
    item?.title?.substring(0, searchText.length).includes(searchText)
  );

  const showEmptyState = filteredCategoryList.length === 0;

  useLayoutEffect(() => {
    navigation.setOptions({
      headerTitle: category,
      headerLeft: () => (
        <Pressable onPress={() => navigation.pop()}>
          <Ionicons name="arrow-back-outline" size={24} color={Colors.white} />
        </Pressable>
      ),
    });
  }, [navigation]);

  const {container, productCardStyle, textInputView} = styles;

  const renderProduct = ({item}: {item: Product}) => {
    return (
      <ProductCard
        title={item.title}
        imageUrl={item.image}
        price={item.price}
        onPress={() =>
          navigation.navigate("ProductDetail", {productId: item.id})
        }
        style={productCardStyle}
      />
    );
  };

  const onChangeSearch = (text: string) => {
    setSearchText(text);
  };

  return (
    <View style={container}>
      <View style={textInputView}>
        <CustomTextInput
          placeholder="Search"
          onChangeText={onChangeSearch}
          value={searchText}
          backgroundColor={Colors.mainNight}
          placeholderTextColor={Colors.mainYellow}
          textColor={Colors.mainYellow}
        />
      </View>

      {showEmptyState ? (
        <EmptyState emptyStateText="We couldn't find any result" />
      ) : (
        <FlatList
          data={filteredCategoryList}
          numColumns={2}
          renderItem={renderProduct}
          showsVerticalScrollIndicator={false}
          keyExtractor={(_, index) => String(index)}
        />
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.mainGrey,
  },
  textInputView: {
    padding: 12,
  },
  productCardStyle: {
    flex: 1,
    margin: 12,
  },
});
