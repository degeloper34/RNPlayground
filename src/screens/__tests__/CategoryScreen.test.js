import renderer from "react-test-renderer";
import React from "react";
import CategoryScreen from "../CategoryScreen";

describe("CategoryScreen", () => {
  it("should render without issues", () => {
    const setOptions = jest.fn();
    const navigate = jest.fn();

    const screen = renderer
      .create(
        <CategoryScreen
          navigation={{
            navigate,
            setOptions,
          }}
          route={{
            params: {
              categoryTitle: "men's clothing",
              category: [
                {
                  id: 1,
                  title:
                    "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
                  category: "men's clothing",
                  description:
                    "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
                  image:
                    "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
                  price: 109.95,
                  rating: {
                    count: 120,
                    rate: 3.9,
                  },
                },
              ],
            },
          }}
        />
      )
      .toJSON();

    expect(screen).toMatchSnapshot();
  });
});
