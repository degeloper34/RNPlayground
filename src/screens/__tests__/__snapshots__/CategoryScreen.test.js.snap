// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryScreen should render without issues 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F0F0F0",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "padding": 12,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "borderRadius": 8,
            "justifyContent": "center",
            "paddingHorizontal": 6,
            "paddingVertical": 12,
          },
          Object {
            "backgroundColor": "#21325E",
          },
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Search"
        placeholderTextColor="#F1D00A"
        rejectResponderTermination={true}
        style={
          Object {
            "color": "#F1D00A",
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "category": "men's clothing",
          "description": "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
          "id": 1,
          "image": "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
          "price": 109.95,
          "rating": Object {
            "count": 120,
            "rate": 3.9,
          },
          "title": "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "backgroundColor": "white",
                  "borderRadius": 8,
                  "height": 381.14285714285717,
                  "overflow": "hidden",
                  "width": 250,
                },
                Object {
                  "flex": 1,
                  "margin": 12,
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "flex": 3,
                }
              }
            >
              <Image
                resizeMode="contain"
                source={
                  Object {
                    "uri": "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
                  }
                }
                style={
                  Object {
                    "flex": 1,
                  }
                }
              />
            </View>
            <View
              style={
                Object {
                  "backgroundColor": "#21325E",
                  "flex": 1,
                  "marginTop": 12,
                  "padding": 3,
                }
              }
            >
              <View
                style={
                  Object {
                    "flex": 2,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  numberOfLines={2}
                  style={
                    Array [
                      undefined,
                      Object {
                        "color": "#F0F0F0",
                        "fontFamily": "sora-medium",
                        "fontSize": 12,
                        "textDecorationLine": "none",
                      },
                    ]
                  }
                >
                  Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
                </Text>
              </View>
              <View
                style={
                  Object {
                    "flex": 1,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  numberOfLines={2}
                  style={
                    Array [
                      undefined,
                      Object {
                        "color": "#F1D00A",
                        "fontFamily": "sora-bold",
                        "fontSize": 12,
                        "textDecorationLine": "none",
                      },
                    ]
                  }
                >
                  $109.95
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
